Preserve default argument text through serialization

jrose-apple opened the pull request:
This allows us to dump it in the generated interface, though it's still not syntax-highlighted. This is necessary for textual module interfaces, but it's also just a longstanding request for Xcode's "Generated Interface" / "Jump to Definition" feature.

[SR-2608](https://bugs.swift.org/browse/SR-2608) / rdar://problem/18675831

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

rintaro commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -3057,8 +3057,11 @@ ModuleFile::getDeclCheckedImpl(DeclID DID, Optional<DeclContext *> ForcedContext
+    if (auto defaultArg = getActualDefaultArgKind(rawDefaultArg)) {
       param->setDefaultArgumentKind(*defaultArg);
+      if (!blobData.empty())
+        param->setDefaultValueStringRepresentation(blobData);
```
Probably I'm missing something. But, where the `blobData` come from? Isn't this `StringRef` backed by `SmallVector<uint64_t, 64> scratch;`? Is it safe to store it in AST nodes as a `StringRef`?

rintaro replied:
Ah, I read the source of `BitstreamReader`, and I understand it now :)

rintaro reviewed (approved):
This is great! Please add this to `CHANGELOG`!

benlangmuir commented on test/SourceKit/CursorInfo/cursor_info.swift:
```diff
@@ -652,13 +652,11 @@ enum E7: String {
+// CHECK73-SAME: = []
+// CHECK73-SAME: = [:]
+// FIXME: should be <syntaxtype.keyword>nil</syntaxtype.keyword>
+// CHECK73-SAME: = nil
```
Do you know why this isn't getting the keyword treatment?  Is it coming through as a "normal" default argument?

jrose-apple replied:
Yeah, any explicitly-written non-magic-literal expression is still "normal". I suppose we could special-case "the single token 'nil'" and see if that works, but it didn't seem too important to me when *other* default arguments are *also* not syntax-highlighted.

benlangmuir replied:
Huh, so are `DefaultArgumentKind::NilLiteral` (and the empty literals too I guess) unused?

jrose-apple replied:
They're used by the importer, since synthesizing instructions and type-checking them would have been a pain. It also mattered back before default arguments were inlinable, because you definitely don't want to call a function just to get `nil`.

jrose-apple replied:
I experimented with the "see if the expression is a NilLiteralExpr" approach and it fell over in Sema. It's probably still doable, but I think I'm just going to leave it for now.

benlangmuir replied:
SGTM

jrose-apple committed:
Preserve default argument text through serialization

This allows us to dump it in the generated interface, though it's
still not syntax-highlighted. This is necessary for textual module
interfaces, but it's also just a longstanding request for Xcode's
"Generated Interface" / "Jump to Definition" feature.

rdar://problem/18675831

jrose-apple commented:
@swift-ci Please smoke test
